sources/stake.move

```diff
@@ -225,62 +243,128 @@ module harvest::stake {
         let stake_scale = math128::pow(10, (coin::decimals<S>() as u128));
         let scale = stake_scale * reward_scale;
 
-        let pool = StakePool<S, R> {
+        let epoch = Epoch {
+            rewards_amount,
+
             reward_per_sec,
             accum_reward: 0,
-            last_updated: current_time,
-            start_timestamp: current_time,
-            end_timestamp,
+
+            start_time: current_time,
+            last_update_time: current_time,
+            end_time: end_timestamp,
+
+            distributed: 0,
+            ended_at: 0,
+        };
+
+        let pool = StakePool<S, R> {
+            current_epoch: 0,
+            epochs: vector[epoch],
+
             stakes: table::new(),
             stake_coins: coin::zero(),
             reward_coins,
             scale,
             total_boosted: 0,
+
             nft_boost_config,
+            whitelist: table_with_length::new(),
 
             emergency_locked: false,
+
             stake_events: account::new_event_handle<StakeEvent>(owner),
             unstake_events: account::new_event_handle<UnstakeEvent>(owner),
             deposit_events: account::new_event_handle<DepositRewardEvent>(owner),
             harvest_events: account::new_event_handle<HarvestEvent>(owner),
             boost_events: account::new_event_handle<BoostEvent>(owner),
             remove_boost_events: account::new_event_handle<RemoveBoostEvent>(owner),
         };
+
+        add_into_whitelist_inner(&mut pool, whitelist);
+
         move_to(owner, pool);
     }
 
     /// Depositing reward coins to specific pool, updates pool duration.
     ///     * `depositor` - rewards depositor account.
     ///     * `pool_addr` - address under which pool are stored.
     ///     * `coins` - R coins which are used in distribution as reward.
-    public fun deposit_reward_coins<S, R>(depositor: &signer, pool_addr: address, coins: Coin<R>) acquires StakePool {
+    ///     * `duration` - new pool life duration.
+    public fun deposit_reward_coins<S, R>(
+        depositor: &signer,
+        pool_addr: address,
+        coins: Coin<R>,
+        duration: u64,
```

Comment:
Can we have also start of epoch?

Reply:
Sounds like a big deal, I can't say for sure

Reply:
if it's too much changes it's ok